### PR TITLE
hw06

### PR DESCRIPTION
**Pull request by evc8p**
Added the CustomExecutorService and CustomWebServer objects

### CODE REVIEW

File: src/main/java/com/evch/rrm/CustomExecutorService.java [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+                        workers.forEach(worker -> worker.interrupt());
+                        workers.clear();
+                        tasks.clear();
+                    }
+                } catch (InterruptedException e) {
+                    e.printStackTrace();
+                }
+            }
+        });
+    }
```
Comment (gneginskiy) [inline src/main/java/com/evch/rrm/CustomExecutorService.java:78]: busy wait is consuming CPU a lot

File: src/main/java/com/evch/rrm/CustomExecutorService.java [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+
+    @Override
+    public boolean awaitTermination(long timeout, TimeUnit unit) {
+        isAwaitTermination = true;
+        lock.lock();
+        long timeoutNs = unit.toNanos(timeout);
+        long time = System.nanoTime(), t = 0L;
+        boolean result = true;
+        try {
+            while (!isTerminated()) {
```
Comment (gneginskiy) [inline src/main/java/com/evch/rrm/CustomExecutorService.java:111]: busy wait w/o sleep / wait ? cpu-expensive

File: src/main/java/com/evch/rrm/CustomExecutorService.java [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+        ces.testPerformanceComparison();
+        ces.testConcurrentExecution();
+        ces.testShutdownBehavior();
+    }
+
+    private void startThreadsExecutor() {
+        Thread.ofVirtual().start(() -> {
+            while (!isTerminated() || isAwaitTermination) {
+                try {
+                    if (workers.size() < maxCorePoolSize && !isShutdownNow) {
```
Comment (gneginskiy) [inline src/main/java/com/evch/rrm/CustomExecutorService.java:53]: in reality there's no purpose in reusing virtual threads. thus, maxCorePoolSize does not make much sense for virtual threads. it's ok to use them as dispensibles.

File: src/main/java/com/evch/rrm/CustomExecutorService.java [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+                        Runnable task = tasks.poll(waitingForNewTasksTimeoutMs, TimeUnit.MILLISECONDS);
+                        if (task != null) {
+                            workers.addLast(useVirtualThreads ? Thread.ofVirtual().start(task) : Thread.ofPlatform().start(task));
```
Comment (gneginskiy) [inline src/main/java/com/evch/rrm/CustomExecutorService.java:56]: race condition: 
workers.addLast, removeIf, getFirst() etc., on LinkedList without locking, while multiple threads potentially can  access it concurrently.

the method itself can be called twice, and will create a race condition rather sooner than later

File: src/main/java/com/evch/rrm/webserver/CustomWebServer.java [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+        } else {
+            send404(clientSocket);
+        }
+        return !(requestEntries.containsKey("connection") && requestEntries.get("connection").contains("close"));
+    }
+
+    private boolean checkRequest(Map<String, String> headers) {
+        return (headers.containsKey("get") && headers.get("http_type").contains("http/1.1")
+                && headers.containsKey("host") && (headers.get("host").contains("localhost:8080") || headers.get("host").contains("localhost:8081"))
+                && headers.containsKey("user-agent") && headers.get("user-agent").contains("mozilla/5.0")
```
Comment (gneginskiy) [inline src/main/java/com/evch/rrm/webserver/CustomWebServer.java:170]: don't get it. we allow only certain requests from certain browsers? 
Why hardcode 1 particular case? if the current host is 8082, then, it should not work?

File: src/main/java/com/evch/rrm/webserver/CustomWebServer.java [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+                            }
+                        }
+                    });
+                } catch (SocketTimeoutException ste) {
+                } catch (IOException e) {
+                    e.printStackTrace();
+                }
+            }
+        });
+    }
```
Comment (gneginskiy) [inline src/main/java/com/evch/rrm/webserver/CustomWebServer.java:107]: try with resources for clientsocket?

**Review by gneginskiy — APPROVED** [review]
looks reasonable overall. good job with keepalives. 

what can be improved: 
- code modularity (try to cover with unit tests, and it'll be obvious that one method is doing a lot of things, maybe too many)
- edge cases handling
